rfd/0053-passwordless-fido2.md

```diff
@@ -0,0 +1,494 @@
+
+Linking for go-libfido2 varies depending on OS - it defaults to
+[static for macOS](
+https://github.com/keys-pub/go-libfido2/blob/master/fido2_static.go) (with an
+embedded libcbor binary) and [dynamic for Linux and Windows](
+https://github.com/keys-pub/go-libfido2/blob/master/fido2_other.go#L6) (with
+various embedded binaries for Windows). We should consider replacing those
+during build, or once again forking go-libfido2 in order to standardize building
+and linking strategies. (Static linking seems to be the best option for `tsh`.)
+
+During development, the libfido2-aware code in Teleport is to be protected by a
+`libfido2` build tag - said tag could be kept in the final product if dealing
+with the library proves to be a hassle.
+
+### Security
+
+Security considerations are largely unchanged in relation to the
+[Passwordless RFD][passwordless rfd].
+
+It's important to note that CLI applications with hardware key access are
+particularly dangerous, as they are not subject to the security measures
+we've come to expect from browsers (in fact, _any_ untrusted application that
+interacts with hardware keys should be used with caution).
+Those concerns are not exclusive to WebAuthn, but are made more delicate in face
+of passwordless credentials.
+
+`tsh` does its best to be well-behaved and mitigates hardware key access
+concerns by being open source, providing signed binaries and its own Cloud
+solution.
+
+(Security considerations for the [browser-based CLI authentication](
+#browser-based-cli-authentication) alternative section are contained within it.)
+
+### UX
+
+UX is discussed throughout the design, but here is a summary of changes:
+
+**Browser-based clients**
+
+The exact passwordless flow is browser-dependent, see the diagram in the
+[browser-based clients](#browser-based-clients) section for a quick reference.
+
+Web UI changes are outside of the scope of the design, but it is likely that the
+user will need the means to choose between "regular" and "passwordless" flows
+before the ceremony starts. A real-world example of such flow may be found at
+Microsoft's https://login.live.com/.
+
+Registration has to make the same decision: does it take a resident key slot to
+allow passwordless? The decision is delegated to the Web UI, but it is likely
+users will want to decide in a case-by-case basis, specially when using hardware
+keys with storage limitations.
+
+**CLI clients (aka `tsh`)**
+
+Similarly to browser-based authentication, `tsh login` needs to know beforehand
+whether to go for the "regular" or "passwordless" flow. The decision needs to
+come from the user, otherwise we risk a needless hardware key prompt (and the
+clunky UX that comes with it).
+
+Example of a login with multiple hardware keys, PIN, and multiple credentials
+(some steps may be skipped in simpler scenarios, see the
+[CLI-native authentication](#cli-native-authentication) section):
+
+```shell
+$ tsh login --proxy=example.com --pwdless
```

Comment:
For FIDO2 I think so. The best we can tell is if there are any hardware keys plugged in the system, from this point on we would have to assume that one of them holds a passwordless key and start the flow. If we guessed wrong there won't be a resident key and we'll fail partway, with I think at least one needless touch, so not that great.

Touch ID makes this a lot nicer, since in that case we can tell if a key exists without user interaction.